### PR TITLE
Update build-containers.yml

### PR DESCRIPTION
**Pull request by Sarthak-Singh04**
I'm using actions/checkout@v4, docker/setup-buildx-action@v3, and actions/cache@v3 to ensure we have the latest features and security updates.

Implemented Docker layer caching

Managed build cache effectively

Fixes: #7509 

# Checklist:

- [x] The code follows the [coding guidelines](https://github.com/omegaup/omegaup/wiki/Coding-guidelines) of omegaUp.
- [x] The tests were executed and all of them passed.
- [x] If you are creating a feature, the new tests were added.
- [x] If the change is large (> 200 lines), this PR was split into various Pull Requests. It's preferred to create one PR for changes in controllers + unit tests in PHPUnit,  and then another Pull Request for UI + tests in Jest, Cypress or both.

### CODE REVIEW

**Comment by pabo99** [comment]
Could you please add an issue indicating why we need this change? After that we can link it with this PR

**Comment by Sarthak-Singh04** [comment]
I just updating our CI/CD setup to make our build process faster and more efficient. We're doing this by using Docker Buildx for better builds, updating our GitHub Actions for the latest features, and improving how we handle caching. This means quicker build times and a more streamlined workflow. It's all about speeding up development and staying up-to-date with the best tools available.

**Comment by pabo99** [comment]
Could we assess the performance after implementing these changes?

File: .github/workflows/build-containers.yml [hunk 1]
```diff
@@ -11,26 +11,44 @@ jobs:
     runs-on: ubuntu-20.04
 
     steps:
-      - uses: actions/checkout@v2
+      - uses: actions/checkout@v4
```
Comment (pabo99) [inline .github/workflows/build-containers.yml:14]: Do we need to update this version in the file `ci.yml` and also in the other `.yml` files?

**Comment by Sarthak-Singh04** [comment]
Yes, we'll track and compare key performance metrics before and after making these changes to evaluate their impact.

File: .github/workflows/build-containers.yml (hunk 1 above)
Comment (Sarthak-Singh04) [reply .github/workflows/build-containers.yml:14]: Yes, it's a good practice to update the checkout action to v4 in all our .yml workflow files. . Updating to v4 across the board also means we can take advantage of the latest features and improvements in the checkout action, potentially enhancing our workflow efficiency and reliability.

**Comment by Sarthak-Singh04** [comment]
> Could we assess the performance after implementing these changes?

you can take reference from this article https://dev.to/ken_mwaura1/optimizing-github-actions-performance-enhance-workflows-with-caching-4hla  how much we optimize the performance

File: .github/workflows/build-containers.yml [hunk 2]
```diff
@@ -11,26 +11,44 @@ jobs:
...
+        with:
+          path: /tmp/.buildx-cache
+          key: ${{ runner.os }}-buildx-${{ github.sha }}
+          restore-keys: |
+            ${{ runner.os }}-buildx-
+
       - name: Build docker containers
         env:
           COMPOSE_DOCKER_CLI_BUILD: 1
-          DOCKER_BUILDKIT: 1
```
Comment (pabo99) [inline .github/workflows/build-containers.yml:24]: Why did you remove this line? It could work with Docker Buildx, right?

Comment (Sarthak-Singh04) [reply .github/workflows/build-containers.yml:24]: we use the buildx actions  for the build process so we don't need to set the Docker_buildkit .

**Review by pabo99 — APPROVED** [review]
LGTM

**Review by heduenas — APPROVED** [review]
